### DISCUSSION TITLE
How to get DatabaseConnection with param in isolate.

### DISCUSSION THREAD

**Discussion by HuangZiquan** [simolus3/drift#2210] [Q&A] [answered]
Hi,
I want to **sync data in background** using [computer.compute](https://pub.dev/packages/computer)(like load balancing), and **notify UI isolate to update**.
```
void setupSqlCipher() {
  open.overrideFor(OperatingSystem.android, () => DynamicLibrary.open('libsqlcipher.so'));
}

Future<DriftIsolate> _createDriftIsolate(String userUid) async {
  final dbFolder = await getApplicationSupportDirectory();
  final path = p.join(dbFolder.path, userUid, "$userUid.sqlite");
  final receivePort = ReceivePort();
  await Isolate.spawn(
    _startBackground,
    _IsolateStartRequest(receivePort.sendPort, path),
  );
  return await receivePort.first as DriftIsolate;
}

void _startBackground(_IsolateStartRequest request) {
  final executor = LazyDatabase(() async {
    setupSqlCipher();
    return NativeDatabase(
      File(request.targetPath),
      setup: (rawDb) {
        rawDb.execute("PRAGMA key = 'password';");
      },
    );
  });
  final driftIsolate = DriftIsolate.inCurrent(
    () => DatabaseConnection(executor),
  );
  request.sendDriftIsolate.send(driftIsolate);
}

class _IsolateStartRequest {
  final SendPort sendDriftIsolate;
  final String targetPath;

  _IsolateStartRequest(this.sendDriftIsolate, this.targetPath);
}

DatabaseConnection createDriftIsolateAndConnect(String userUid) {
  return DatabaseConnection.delayed(Future.sync(() async {
    final isolate = await _createDriftIsolate(userUid);
    return await isolate.connect(singleClientMode: true);
  }));
}

AppDatabase openDB(String userUid) {
  return AppDatabase.connect(createDriftIsolateAndConnect(userUid));
}
```

How to get DatabaseConnection with param in isolate, and notify UI isolate to update.
```
await computer.compute(
  syncDataInBackgroundIsolate,
  param: {"userUid":userUid},
);

syncDataInBackgroundIsolate(Map map) async {
    final userUid = map["userUid"] as String;

    //my problem is here
    MoorIsolate isolate = await MoorIsolate.spawn(createDriftIsolateAndConnect);
    DatabaseConnection connection = await isolate.connect();
    final AppDatabase db = AppDatabase.connect(connection);

   ... sync data ...

    db.close();
    isolate.shutdownAll();
}
```

By the way, am I doing it right?

**Comment by simolus3** [comment, marked as answer]
In this case, I wouldn't spawn an additional drift isolate in `syncDataInBackgroundIsolate` like you're doing it. Instead, spawn a drift isolate _once_ and use it for both the UI and the background compute unit.

```
await computer.compute(
  syncDataInBackgroundIsolate,
  param: {"userUid":userUid, 'drift': yourDriftIsolate},
);

syncDataInBackgroundIsolate(Map map) async {
    final userUid = map["userUid"] as String;
    final connection = await (map['drift'] as DriftIsolate).connect();

    final AppDatabase db = AppDatabase.connect(connection);

   ... sync data ...

    db.close();
}
```

If you're using the same `DriftIsolate`, query streams get synchronized automatically. So you don't need to manually do the synchronization in that case.

**Reply by HuangZiquan** [reply]
Hi,
I use the same `DriftIsolate` for UI and background.
I use `watch` to listen user table in UI.
When I sync User data in the background, I will not get user data in UI.

```
Future<void> startSyncUsers() async {
    final List<User> users = Api.getUsers();

    final executor = Executor(concurrency: getNumberOfProcessors());
    for (var user in users ) {
      executor.scheduleTask(() async {
        await computer.compute(syncUserInBackground, param: {
            'drift': myDriftIsolate,
        });
      });
    }
    await executor.join(withWaiting: true);
    await executor.close();
}

syncUserInBackground(Map map) async {
    final connection = await (map['drift'] as DriftIsolate).connect();
    final AppDatabase db = AppDatabase.connect(connection);
    
    await db.transaction(() async {
        ... call user information api or do something...
    });

    await  db.close();
}
```

It seems to be operating the DB in the background and affecting the listening behavior in UI.

**Reply by simolus3** [reply]
Do you have perhaps a full example of the setup logic that I could check out to reproduce this issue? Your code looks correct to me know.

You could also enable `isolateDebugLog` in `connect()` on both the UI and the background service. Drift should print a notification for lots of things then, including something about receiving a tables-updated notification. Does something like that happen in your case or is that message not there?

**Reply by HuangZiquan** [reply]
I found out that my problem is [this](https://github.com/simolus3/drift/issues/320).
When I continue to write data in the background, the DB is locked and the query in the foreground is slow.
Is there a way to use concurrency?

**Reply by simolus3** [reply]
Do you really get a "database is locked" exception? That would be a pretty good indication that you indeed have _two_ opened database instances instead of one.

>  Is there a way to use concurrency?

You can use a `MultiExecutor` as described in the linked issue to synchronize writes but keep reads local for performance.
